core/data.go

```diff
@@ -52,8 +53,111 @@ const (
 	// in protobuf, UNKNOWN as 0 is a convention).
 	GobBundleEncodingFormat   BundleEncodingFormat = 0
 	GnarkBundleEncodingFormat BundleEncodingFormat = 1
+
+	// Similar to bundle encoding format, this describes the encoding format of chunks.
+	// The difference is ChunkEncodingFormat is just about chunks, whereas BundleEncodingFormat
+	// is also about how multiple chunks of the same bundle are packed into a single byte array.
+	GobChunkEncodingFormat   ChunkEncodingFormat = 0
+	GnarkChunkEncodingFormat ChunkEncodingFormat = 1
 )
 
+type ChunksData struct {
+	// Chunks is the encoded bytes of the chunks.
+	Chunks [][]byte
+	// Format describes how the bytes of the chunks are encoded.
+	Format ChunkEncodingFormat
+	// The number of symbols in each chunk.
+	// Note each chunk of the same blob will always have the same number of symbols.
+	ChunkLen int
+}
+
+func (cd *ChunksData) Size() uint64 {
+	if len(cd.Chunks) == 0 {
+		return 0
+	}
+	// GnarkChunkEncoding will create chunks of equal size.
+	if cd.Format == GnarkChunkEncodingFormat {
+		return uint64(len(cd.Chunks)) * uint64(len(cd.Chunks[0]))
+	}
+	// GobChunkEncoding can create chunks of different sizes.
+	size := uint64(0)
+	for _, c := range cd.Chunks {
+		size += uint64(len(c))
+	}
+	return size
+}
+
+func (cd *ChunksData) FlattenToBundle() ([]byte, error) {
+	// Only Gnark coded chunks are dispersed as a byte array.
+	// Gob coded chunks are not flattened.
+	if cd.Format != GnarkChunkEncodingFormat {
+		return nil, fmt.Errorf("unsupported chunk encoding format to flatten: %v", cd.Format)
+	}
+	result := make([]byte, cd.Size()+8)
+	buf := result
+	metadata := (uint64(cd.Format) << (NumBundleHeaderBits - NumBundleEncodingFormatBits)) | uint64(cd.ChunkLen)
+	binary.LittleEndian.PutUint64(buf, metadata)
+	buf = buf[8:]
+	for _, c := range cd.Chunks {
+		if len(c) != len(cd.Chunks[0]) {
+			return nil, errors.New("all chunks must be of same size")
+		}
+		copy(buf, c)
+		buf = buf[len(c):]
+	}
+	return result, nil
+}
+
+func (cd *ChunksData) ToGobFormat() (*ChunksData, error) {
+	if cd.Format == GobChunkEncodingFormat {
+		return cd, nil
+	}
+	if cd.Format != GnarkChunkEncodingFormat {
+		return nil, fmt.Errorf("unsupported chunk encoding format: %d", cd.Format)
+	}
+	gobChunks := make([][]byte, 0, len(cd.Chunks))
```

Comment:
So for a while, serialization will still be done in batcher?

Reply:
I guess the plan is to gob encode in the encoder.

Reply:
Yes, the serialization will happen only once, which will be at Encoder. That'll in a separate PR.